src/hotspot/share/opto/runtime.cpp

```diff
@@ -132,13 +125,62 @@ static bool check_compiled_frame(JavaThread* thread) {
 }
 #endif // ASSERT
 
-
+/*
 #define gen(env, var, type_func_gen, c_func, fancy_jump, pass_tls, return_pc) \
   var = generate_stub(env, type_func_gen, CAST_FROM_FN_PTR(address, c_func), #var, fancy_jump, pass_tls, return_pc); \
   if (var == nullptr) { return false; }
+*/
+
+#define GEN_OPTO_BLOB(name, type)                    \
+  generate_ ## name ## _blob();
+
+// a few helper macros to conjure up generate_stub call arguments
+#define OPTO_STUB_FIELD_NAME(name) _ ## name ## _Java
+#define OPTO_STUB_TYPEFUNC(name) name ## _Type
+#define OPTO_STUB_C_FUNC(name) CAST_FROM_FN_PTR(address, name ## _C)
+#define OPTO_STUB_NAME(name) stub_name(OptoStubId::name ## _id)
+
+// Almost all the C functions targeted from the generated stubs are
+// implemented locally to OptoRuntime with names that can be generated
+// from the stub name by appending suffix '_C'. However, in two cases
+// a common target method also needs to be called from shared runtime
+// stubs. In these two cases the opto stubs rely on method
+// imlementations defined in class SharedRuntime. The following
+// defines temporarily rebind the generated names to reference the
+// relevant implementations.
+
```

Comment:
I am not 100% happy about using defines to finesse this problem of common C targets.

One alternative here is to define methods local to class OptoRuntime which fit the generator naming convention and have them forward the call to the SharedRuntime methods. n.b. I used (local) method forwarding to allow blobs to share common typefunc providers.

Another alternative is to declare the C target as a parameter to the opto blob declaration macro. That's more flexible but in almost all cases it repeats information already present and makes understanding and updating the declarations more complex.

Reply:
This is indeed confusing even with comment. 
I prefer your first suggestion: "define methods local to class OptoRuntime which fit the generator naming convention and have them forward the call to the SharedRuntime methods"

Reply:
Done.